src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -22,6 +22,181 @@ namespace System
+        private static ReadOnlySpan<T> TrimImpl<T>(ref ReadOnlySpan<T> span, T trimElement, bool trimStart, bool trimEnd)
+            where T : IEquatable<T>
+        {
+            if (span.IsEmpty)
+                return span;
+
+            int start = 0;
+            int end = span.Length - 1;
+
+            if (trimElement == null)
+            {
+                if (trimStart)
+                {
+                    for (; start < span.Length; start++)
+                    {
+                        if (span[start] != null)
+                            break;
+                    }
+                }
+
+                if (trimEnd)
+                {
+                    for (; end >= start; end--)
+                    {
+                        if (span[end] != null)
+                            break;
+                    }
+                }
+            }
+            else
+            {
+                if (trimStart)
+                {
+                    for (; start < span.Length; start++)
+                    {
+                        if (span[start] != null && !span[start].Equals(trimElement))
+                            break;
+                    }
+                }
+
+                if (trimEnd)
+                {
+                    for (; end >= start; end--)
+                    {
+                        if (span[end] != null && !span[end].Equals(trimElement))
+                            break;
+                    }
+                }
+            }
+
+            return span.Slice(start, end - start + 1);
+        }
+
+        /// <summary>
+        /// Removes all leading and trailing occurrences of a set of elements specified
+        /// in a readonly span from the span.
+        /// </summary>
+        /// <param name="span">The source span from which the elements are removed.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <remarks>If <paramref name="trimElements"/> is empty, the span is returned unaltered.</remarks>
+        public static ReadOnlySpan<T> Trim<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+            => span.TrimStart(trimElements).TrimEnd(trimElements);
+
+        /// <summary>
+        /// Removes all leading occurrences of a set of elements specified
+        /// in a readonly span from the span.
+        /// </summary>
+        /// <param name="span">The source span from which the elements are removed.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <remarks>If <paramref name="trimElements"/> is empty, the span is returned unaltered.</remarks>
+        public static ReadOnlySpan<T> TrimStart<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            if (span.IsEmpty || trimElements.IsEmpty)
```

Comment:
This guard is not present in the existing `char` and `byte` overloads. Should it be?